src/common/utxobased/engine/makeUtxoEngineState.ts

```diff
@@ -240,27 +307,256 @@ const setLookAhead = async (args: SetLookAheadArgs): Promise<void> => {
+    (await processor.fetchAddressByScriptPubkey(scriptPubkey)) ?? {}
+  transactions.set(address, {
+    fetching: false,
+    path: {
+      format,
+      branch
+    },
+    page: 1, // Page starts on 1
+    networkQueryVal
+  })
+}
+
+interface NextTaskArgs extends CommonArgs {
+  blockBook: BlockBook
+}
+
+export const pickNextTask = async (
+  args: NextTaskArgs
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+): Promise<WsTask<any> | undefined> => {
+  const { addressesToWatch, taskCache, blockBook } = args
+
+  const {
+    addressSubscribeCache,
+    utxosCache,
+    rawUtxosCache,
+    processedUtxosCache,
+    transactionsCache,
+    updateTransactionsCache
+  } = taskCache
+
+  console.log({
+    addressSubscribeCache: addressSubscribeCache.size,
+    utxosCache: utxosCache.size,
+    rawUtxosCache: rawUtxosCache.size,
+    processedUtxosCache: processedUtxosCache.size,
+    transactionsCache: transactionsCache.size,
+    updateTransactionsCache: updateTransactionsCache.size
+  })
+
+  // first check if blocks are already being watched
+  if (!taskCache.blockWatching) {
+    taskCache.blockWatching = true
+    const deferred = new Deferred<unknown>()
+    deferred.promise.catch(() => {
+      taskCache.blockWatching = false
+    })
+
+    blockBook.watchBlocks(
+      async () => await onNewBlock({ ...args, blockBook }),
+      deferred
+    )
+  }
+
+  // Loop to process addresses to utxos
+  for (const [address, state] of utxosCache) {
+    // Check if we need to fetch address UTXOs
+    if (!state.fetching) {
+      state.fetching = true
+
+      utxosCache.delete(address)
+
+      // Fetch and process address UTXOs
+      const wsTask = await processAddressUtxos({
+        ...args,
+        ...state,
+        address
+      })
+      return wsTask
+    }
+  }
+
+  // Check if there are any addresses pending to be subscribed
+  if (addressSubscribeCache.size > 0 && !taskCache.addressWatching) {
+    // Loop each address that needs to be subscribed
+    for (const [address, path] of addressSubscribeCache) {
+      // Add address in the cache to the set of addresses to watch
+      addressesToWatch.set(address, path)
+
+      if (path != null) {
+        // Add the newly watched addresses to the UTXO cache
+        utxosCache.set(address, {
+          fetching: false,
+          path
+        })
+        await addToTransactionCache(
+          args,
+          address,
+          path.format,
+          path.branch,
+          transactionsCache
+        )
+      }
+    }
+
+    taskCache.addressWatching = true
+
+    const deferred = new Deferred<unknown>()
+    deferred.promise
+      .then(() => {
+        addressSubscribeCache.clear()
+      })
+      .catch(() => {
+        taskCache.addressWatching = false
+      })
+    blockBook.watchAddresses(
+      Array.from(addressesToWatch.keys()),
+      (response: INewTransactionResponse) => {
+        const path = addressesToWatch.get(response.address)
```

Comment:
Check if addresses have a path when adding to addressesToWatch/addressSubscribe

Reply:
This seems to already be the case in both instances setting the addressSubscribeCache.  SetLookAhead uses walletTools to get the address by path, processPathAddresses, gets the addresses from getNumAddressesFromPathPartition.